### PR TITLE
ESROGUE-527: Allow Hub class to automatically split large transactions

### PR DESCRIPTION
**Pull request by bengineerd**
### Description
This change allows the `Hub` class to automatically split large transactions into smaller transactions. 

This is particularly useful when large `RemoteVariables` are accessed over SRPv3, where the max transaction size is 4096 bytes. Previously, such variables had to be split into multiple variables in the `Device` subclass.

#### Hub Splitting
The `Hub.reqTransaction()` method now checks the max transactions size allowed by the downstream `Slave`. If the incoming transaction fits, it is passed through just as before. If it is too big, then it will be split into multiple subtransactions. 

`Hub.doMaxAccess()` now returns MAX_UINT32, since can handle the largest possible transactions by splitting them.

#### Subtransactions
The `Transaction` class has been enhanced with the concept of "subtransactions". 
 - A `memory.Slave` can create multiple subtransactions based on a `Transaction` received via `reqTransaction()`. 
 - A `Transaction` holds a map of its subtransactions, and each subtransaction holds a pointer back to the parent. 
 - A subtransaction can also be split into further subtransactions by a `Slave` further down the memory chain.
 - `Transaction.done()`
   - Checks if there is a parent transaction
   - Removes itself from the parent subtransaction map.
   - Calls `done()` on the parent Transaction if it is the last subtransaction to complete.
 - `Transaction.error()` 
   - Checks if there is a parent transaction 
   - Removes itself from the parent subtransaction map 
   - Calls `error()` on the parent `Transaction`.
 - `Master.waitTransaction()` is only called on the original parent transaction.
   - Subtransactions are not added to the `tranMap_` in their `Hub` that generated them.

### CODE REVIEW

**Comment by mskoufis** [comment]
In regards to the comment that a subtransaction can also be slit into further subtransactions by a Slave further down the memory chain, when will that happen?  If we split a transaction once to conform with the memory size limitation according to the protocol for that slave, then --even if we have a chain of hubs-- there won't be a need to split a subtransaction again as the size limit won't be violated.  The subtransaction will be simply forwarded to the next level until it reaches the memory slave.  I may be misunderstanding this comment or the use case scenario.

**Comment by mskoufis** [comment]
Also, in general, what happens when a transaction fails?  Regardless of whether that transaction is a parent or a single independent transaction.  Is there a mechanism in place to re-issue that transaction?

**Comment by bengineerd** [comment]
Subtransactions might also be necessary for things like an I2C bus. 
An I2C write might need to
 - Set the chip I2C address
 - Write the I2C register address
 - Write the I2C data
 - Poll for the I2C bus transaction to be complete after each write above.

Now in practice, such things rarely also require large transaction splitting. But architecturally it's good to support, and the code does support it as it is.

When a transaction fails, it is up to the higher level code that issued the transaction to decide what to do. For some devices, like I2C, we expect occasional failures so we reissue a new transaction with the same parameters.

**Comment by mskoufis** [comment]
@bengineerd  Is exposing the subtransaction API to Python required for this task or should we create a new task for this?

**Comment by bengineerd** [comment]
I think this enhancement will allow us to remove `Device._rawTxnChunker()` and `MemoryDevice._txnChunker()`.
The transaction splitting that these do is now automatically done by the Hub.

**Comment by mskoufis** [comment]
My comments below.

1.  I don’t quite understand why reqMaxAccess() was replaced by 0xFFFFFFFF in rim::Hub::doMaxAccess().  Can you explain?
2. Although it's fine the way it is, the block 

```
if (tran->size() >= ((i+1) * maxAccess)) {
    subTran->size_ = maxAccess;
} else {
    subTran->size_ = tran->size() % maxAccess;
}

```
would be easier for me to understand as shown below.  Just a personal preference but it works either way.

```
if (i < numberOfTransactions - 1)
  subTran->size_ = maxAccess;
else
  subTran->size_ = tran->size() % maxAccess
```

**Review by mskoufis — APPROVED** [review]

**Review by mskoufis — APPROVED** [review]

File: src/rogue/interfaces/memory/Transaction.cpp [hunk 1]
```diff
@@ -139,6 +174,20 @@ void rim::Transaction::errorPy(std::string error) {
...
+   if (isSubTransaction_)  {
+      // Get a shared_ptr to the parent transaction
+      rim::TransactionPtr parentTran = parentTransaction_.lock();
+      if (parentTran) {
+         // Remove own ID from parent subtransaction map
+        parentTran->subTranMap_.erase(id_);
+
+         // If this is the last sub-transaction, notify parent transaction it is all done
+         if (parentTran->subTranMap_.empty() and parentTran->doneCreatingSubTransactions_)
+           parentTran->error("Transaction error. Subtransaction failed.");
```
Comment (slacrherbst) [inline src/rogue/interfaces/memory/Transaction.cpp:188]: I think we should check to see if an error already exists in the transaction and either append the error or don't replace the existing error?

File: src/rogue/interfaces/memory/Transaction.cpp [hunk 2]
```diff
@@ -128,6 +149,20 @@ void rim::Transaction::done() {
...
+   // If applicable, notify parent transaction about completion of a sub-transaction 
+   if (isSubTransaction_)  {
+      // Get a shared_ptr to the parent transaction
+      rim::TransactionPtr parentTran = this->parentTransaction_.lock();
+      if (parentTran) {
+         // Remove own ID from parent subtransaction map
+        parentTran->subTranMap_.erase(id_);
+
+         // If this is the last sub-transaction, notify parent transaction it is all done
+         if (parentTran->subTranMap_.empty() and parentTran->doneCreatingSubTransactions_)
```
[the remark this answers is not in the document]
Comment (slacrherbst) [reply src/rogue/interfaces/memory/Transaction.cpp:162]: I think we need to get a lock to the parent transaction before modifying it.

[the remark this answers is not in the document]
Comment (bengineerd) [reply src/rogue/interfaces/memory/Transaction.cpp:162]: The thought was that maybe custom Hub subclasses would need to do something where some subtransactions need to be forwarded before all are created. This API then allows for that to happen.

File: src/rogue/interfaces/memory/Transaction.cpp (hunk 1 above)
Comment (bengineerd) [reply src/rogue/interfaces/memory/Transaction.cpp:188]: We do get a lock on the parent at line 181, no?

**Comment by bengineerd** [comment]
@slacrherbst I addressed some of your review comments in https://github.com/slaclab/rogue/pull/827/commits/7ee00e799c79994b8b8e3c4d9ecdfbd485e00326.

**Comment by bengineerd** [comment]
I've been testing this. One issue is the way that sub-transaction errors are reported. When doing a large write/verify that gets broken into subtransactions, and one of the subtransaction writes fails, it gets reported as a verify error.

```
1650669815.349437:pyrogue.memory.Transaction: Created transaction type=2 id=7, address=0x00000000c0800000, size=16000
1650669815.349449:pyrogue.memory.Hub: Splitting transaction 7 into 4 subtransactions
1650669815.349456:pyrogue.memory.Transaction: Created subTransaction id=8, parent=7
1650669815.349460:pyrogue.memory.Hub: Created subTransaction 8, parent=7, iter=d02d3c0, size=4096, address=c0800000
1650669815.349463:pyrogue.memory.Transaction: Created subTransaction id=9, parent=7
1650669815.349466:pyrogue.memory.Hub: Created subTransaction 9, parent=7, iter=d02e3c0, size=4096, address=c0801000
1650669815.349474:pyrogue.memory.Transaction: Created subTransaction id=10, parent=7
1650669815.349478:pyrogue.memory.Hub: Created subTransaction 10, parent=7, iter=d02f3c0, size=4096, address=c0802000
1650669815.349481:pyrogue.memory.Transaction: Created subTransaction id=11, parent=7
1650669815.349485:pyrogue.memory.Hub: Created subTransaction 11, parent=7, iter=d0303c0, size=3712, address=c0803000
1650669815.349499:pyrogue.SrpV3: Send frame for id=8, addr 0xc0800000. Size=4096, type=2
1650669815.349504:pyrogue.SrpV3: Send frame for id=8, header: 0x0a000103 0x00000008 0xc0800000 0x00000000 0x00000fff
1650669815.349526:pyrogue.SrpV3: Send frame for id=9, addr 0xc0801000. Size=4096, type=2
1650669815.349532:pyrogue.SrpV3: Send frame for id=9, header: 0x0a000103 0x00000009 0xc0801000 0x00000000 0x00000fff
1650669815.349538:pyrogue.SrpV3: Send frame for id=10, addr 0xc0802000. Size=4096, type=2
1650669815.349541:pyrogue.SrpV3: Send frame for id=10, header: 0x0a000103 0x0000000a 0xc0802000 0x00000000 0x00000fff
1650669815.349547:pyrogue.SrpV3: Send frame for id=11, addr 0xc0803000. Size=3712, type=2
1650669815.349550:pyrogue.SrpV3: Send frame for id=11, header: 0x0a000103 0x0000000b 0xc0803000 0x00000000 0x00000e7f
1650669852.202327:pyrogue.SrpV3: Got frame id=8, header: 0x0a000103 0x00000008 0xc0800000 0x00000000 0x00000fff tail: 0x00000000
1650669852.202345:pyrogue.memory.Transaction: Transaction done. type=2 id=8, address=0x00000000c0800000, size=4096
1650669876.671253:pyrogue.SrpV3: Got frame id=9, header: 0x0a000103 0x00000009 0xc0801000 0x00000000 0x00000fff tail: 0x00000000
1650669876.671271:pyrogue.memory.Transaction: Transaction timer refresh! Possible slow link! type=2 id=10, address=0x00000000c0802000, size=4096
1650669876.671275:pyrogue.memory.Transaction: Transaction timer refresh! Possible slow link! type=2 id=11, address=0x00000000c0803000, size=3712
1650669876.671283:pyrogue.memory.Transaction: Transaction done. type=2 id=9, address=0x00000000c0801000, size=4096
1650669876.713401:pyrogue.SrpV3: Got frame id=10, header: 0x0a000103 0x0000000a 0xc0802000 0x00000000 0x00000fff tail: 0x00000003
1650669876.713413:pyrogue.memory.Transaction: Transaction timer refresh! Possible slow link! type=2 id=11, address=0x00000000c0803000, size=3712
1650669876.713429:pyrogue.memory.Transaction: Transaction error. type=2 id=10, address=0x00000000c0802000, size=4096, error=Non zero status message returned on fpga register bus in hardware: 0x3
1650669876.713434:pyrogue.SrpV3: Error detected for ID id=10, tail=0x00000003
1650669899.697283:pyrogue.SrpV3: Got frame id=11, header: 0x0a000103 0x0000000b 0xc0803000 0x00000000 0x00000e7f tail: 0x00000000
1650669899.697301:pyrogue.memory.Transaction: Transaction done. type=2 id=11, address=0x00000000c0803000, size=3712
1650669899.697308:pyrogue.memory.Transaction: Transaction done. type=2 id=7, address=0x00000000c0800000, size=16000
1650669899.697351:pyrogue.memory.Transaction: Created transaction type=4 id=12, address=0x00000000c0800000, size=16000
1650669899.697379:pyrogue.memory.Hub: Splitting transaction 12 into 4 subtransactions
1650669899.697387:pyrogue.memory.Transaction: Created subTransaction id=13, parent=12
1650669899.697393:pyrogue.memory.Hub: Created subTransaction 13, parent=12, iter=d0313d0, size=4096, address=c0800000
1650669899.697399:pyrogue.memory.Transaction: Created subTransaction id=14, parent=12
1650669899.697405:pyrogue.memory.Hub: Created subTransaction 14, parent=12, iter=d0323d0, size=4096, address=c0801000
1650669899.697410:pyrogue.memory.Transaction: Created subTransaction id=15, parent=12
1650669899.697415:pyrogue.memory.Hub: Created subTransaction 15, parent=12, iter=d0333d0, size=4096, address=c0802000
1650669899.697420:pyrogue.memory.Transaction: Created subTransaction id=16, parent=12
1650669899.697424:pyrogue.memory.Hub: Created subTransaction 16, parent=12, iter=d0343d0, size=3712, address=c0803000
1650669899.697433:pyrogue.SrpV3: Send frame for id=13, addr 0xc0800000. Size=4096, type=4
1650669899.697438:pyrogue.SrpV3: Send frame for id=13, header: 0x0a000003 0x0000000d 0xc0800000 0x00000000 0x00000fff
1650669899.697457:pyrogue.SrpV3: Send frame for id=14, addr 0xc0801000. Size=4096, type=4
1650669899.697462:pyrogue.SrpV3: Send frame for id=14, header: 0x0a000003 0x0000000e 0xc0801000 0x00000000 0x00000fff
1650669899.697469:pyrogue.SrpV3: Send frame for id=15, addr 0xc0802000. Size=4096, type=4
1650669899.697474:pyrogue.SrpV3: Send frame for id=15, header: 0x0a000003 0x0000000f 0xc0802000 0x00000000 0x00000fff
1650669899.697480:pyrogue.SrpV3: Send frame for id=16, addr 0xc0803000. Size=3712, type=4
1650669899.697484:pyrogue.SrpV3: Send frame for id=16, header: 0x0a000003 0x00000010 0xc0803000 0x00000000 0x00000e7f
1650669936.199611:pyrogue.SrpV3: Got frame id=13, header: 0x0a000003 0x0000000d 0xc0800000 0x00000000 0x00000fff tail: 0x00000000
1650669936.199636:pyrogue.memory.Transaction: Transaction done. type=4 id=13, address=0x00000000c0800000, size=4096
1650669966.004852:pyrogue.SrpV3: Got frame id=14, header: 0x0a000003 0x0000000e 0xc0801000 0x00000000 0x00000fff tail: 0x00000000
1650669966.004871:pyrogue.memory.Transaction: Transaction timer refresh! Possible slow link! type=4 id=15, address=0x00000000c0802000, size=4096
1650669966.004875:pyrogue.memory.Transaction: Transaction timer refresh! Possible slow link! type=4 id=16, address=0x00000000c0803000, size=3712
1650669966.004882:pyrogue.memory.Transaction: Transaction done. type=4 id=14, address=0x00000000c0801000, size=4096
1650669966.040070:pyrogue.SrpV3: Got frame id=15, header: 0x0a000003 0x0000000f 0xc0802000 0x00000000 0x00000fff tail: 0x00000003
1650669966.040083:pyrogue.memory.Transaction: Transaction timer refresh! Possible slow link! type=4 id=16, address=0x00000000c0803000, size=3712
1650669966.040093:pyrogue.memory.Transaction: Transaction error. type=4 id=15, address=0x00000000c0802000, size=4096, error=Non zero status message returned on fpga register bus in hardware: 0x3
1650669966.040098:pyrogue.SrpV3: Error detected for ID id=15, tail=0x00000003
1650669992.092515:pyrogue.SrpV3: Got frame id=16, header: 0x0a000003 0x00000010 0xc0803000 0x00000000 0x00000e7f tail: 0x00000000
1650669992.092536:pyrogue.memory.Transaction: Transaction done. type=4 id=16, address=0x00000000c0803000, size=3712
1650669992.092545:pyrogue.memory.Transaction: Transaction done. type=4 id=12, address=0x00000000c0800000, size=16000
ERROR:pyrogue.Variable.RemoteVariable.GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam:Block::checkTransaction: General Error: Verify error for block GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam with address 0xc0800000. Byte: 8192. Got: 0x00, Exp: 0x97, Mask: 0xff
Traceback (most recent call last):
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Variable.py", line 794, in set
    self._parent.checkBlocks(recurse=False, variable=self)
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Device.py", line 383, in checkBlocks
    pr.checkTransaction(variable._block, **kwargs)
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Block.py", line 30, in checkTransaction
    block._checkTransaction()
rogue.GeneralError: Block::checkTransaction: General Error: Verify error for block GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam with address 0xc0800000. Byte: 8192. Got: 0x00, Exp: 0x97, Mask: 0xff
ERROR:pyrogue.Variable.RemoteVariable.GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam:Error setting value '[         0    1073741    2147483 ... 4291746069 4292819811 4293893553]' to variable 'GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam' with type uint32(4096,). Exception=Block::checkTransaction: General Error: Verify error for block GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam with address 0xc0800000. Byte: 8192. Got: 0x00, Exp: 0x97, Mask: 0xff
ERROR:pyrogue.Command.BaseCommand.GroupRoot.Group.HardwareGroup.ColumnBoard[0].SetTestRam:Block::checkTransaction: General Error: Verify error for block GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam with address 0xc0800000. Byte: 8192. Got: 0x00, Exp: 0x97, Mask: 0xff
Traceback (most recent call last):
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Command.py", line 115, in _doFunc
    ret = self._functionWrap(function=self._function, root=self.root, dev=self.parent, cmd=self, arg=arg)
  File "<string>", line 1, in <lambda>
  File "./../../firmware/python/warm_tdm/_ColumnModule.py", line 75, in SetTestRam
    self.TestRam.set(value=ram, index=offset, write=True)
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Variable.py", line 799, in set
    raise e
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Variable.py", line 794, in set
    self._parent.checkBlocks(recurse=False, variable=self)
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Device.py", line 383, in checkBlocks
    pr.checkTransaction(variable._block, **kwargs)
  File "/u1/bareese/anaconda3/envs/rogue_build2/lib/python3.9/site-packages/pyrogue/_Block.py", line 30, in checkTransaction
    block._checkTransaction()
rogue.GeneralError: Block::checkTransaction: General Error: Verify error for block GroupRoot.Group.HardwareGroup.ColumnBoard[0].TestRam with address 0xc0800000. Byte: 8192. Got: 0x00, Exp: 0x97, Mask: 0xff

```

**Comment by slacrherbst** [comment]
Yes but you will note there was a previous error. Sadly I think in these cases you have to read back through the list a bit to see the root cause.

**Pull request merged**